WIP: Move stm32 common board logic to stm32/common/

v01d opened the pull request:
## Summary

This is a proposal (to get feedback before continuing) to start moving common board logic for stm32 boards to the `boards/arm/stm32/common` directory. I've started with the BMP180 logic since it is an easy picking but I've been looking into different boards under stm32 and I've found a lot of unnecessary repetition of driver initialization logic (this will probably be true for many other board families). Actually, I started to see divergence in implementations (one had a "missing logic", the other one had been updated) so going forward in this way will avoid the problem.
I thought about adding "do not merge" to this but in itself this could be merged, but it is only the starting point. There are many more cases like this under stm32/ that I can migrate. But before doing such a lengthy task I wanted to get confirmation that I'm on a good path. If you agree on this I can continue pushing here and add "WIP" to the PR.

## Impact

This affects three boards: stm32f103minimum, stm32f4discovery and olimex-stm32-e407. In particular, it affects BMP180 sensor initialization. No functional change is expected.

## Testing

I've built three boards with and without BMP180 support.

patacongo commented:
This looks like an appropriate use of a common directory (it could be a drivers directory like cxd56xx).  Common driver initialization, in general, is a good decision.

Gut we do have to be careful about what we think is common.  Just because two files are the same or similar, does not make them "common" in the architectural sense:

- They must not violate the same partitioning used for other architecutes and boards
- They must not eliminate beneficial architectural hooks that may be need in the future.

Other people have started this in the past and got it complete wrong.  There are attempts to move the stm32_autoleds.c file into common code just become some of there were the same.  Bad Idea!  LEDs are inherently board specific.

There was even an attempt at one point to move the board-specific header file from the board/src directory into a common directory.  A VERY bad ida!

So I think that what you are doing is good and beneficial, we also need to be careful about what we thing of as common logic.  Let's not break the architectural model!  Driver initialization is a good start.  USB host host support is rather complex and very similar to all boards as well.  Things like that would be greate candidates.

v01d commented:
Ok, I'll continue with these simpler devices and eventually move on to more complex cases. Will push here again if you agree.

patacongo commented:
> 
> 
> Ok, I'll continue with these simpler devices and eventually move on to more complex cases. Will push here again if you agree.

I think that common device initialize for drivers under the drivers/ directory is always a good candidate for making common.  The things that need to stay board-specific are those things which are inherently unique to a board (like crystals, buttons, leds) and files that are port of the board architecture/infrastructure (like boot, board header file, bringup, appinit, etc.)

v01d commented:
I'm not sure how to handle license headers now with Apache. I would be probably moving around existing files (currently with BSD header) and modifying them. Should I change the header into Apache? Should I add a second header?

patacongo commented:
> I'm not sure how to handle license headers now with Apache. I would be probably moving around existing files (currently with BSD header) and modifying them. Should I change the header into Apache? Should I add a second header?

All new files requires an Apache 2.0 header.  You can always just move an existing file (which is still not new) and edit it as necessary without changing the header.  If you have multiple choices for which version to common into the common version, select one that already has an Apache 2.0 license or, there is none, select one that I am the sole Copyright holder in the BSD header.

If I am the ONLY copyright holder holder then you can free to replace my BSD copyright header with the Apache 2.0 header (use the one from sched/sched/sched_getcpu.c so that we keep the same line width).

The worst case is that there is no source for the common code that has an Apache license or for which I am the sole Copyright holder.  In that case, I get confused.  But I think we have to copy the BSD header with no changes.  To me that is ugly... Engineers hate inconsistencies.  But I think that is what we have to do.

v01d commented:
I've just moved several other device handling logic into the common folder. I'm moving slowly, it is a big task. For now ignore style errors, since I'm actually having to fix old non compliant files in the process. 
Please review these new cases as some where not as trivial as the others.

I think many could be even be moved into a higher common point since they could actually be applicable to other stm32 variants, but that is left for future work. 
I also think that it would be useful to be able to configure some things via menuconfig (such as I2C/SPI port) for a sensor, which boards usually hardcode. Maybe the common/ dir could expose this via a Kconfig and boards could optionally choose to fix it via own Kconfig entry for board-mounted sensors where the port assignation is fixed.

v01d commented:
Also, I will probably organize files under common/src/ in different subdirectories according to device family. For now I'm trying to make things a bit simpler.

v01d commented:
I realized I must now transition ALL boards under stm32 (not only the ones I modify) to have a Make.defs in src/ instead of a Makefile as the common/ dir's Makefile is now used. It's a bit confusing why this is done so. Couldn't the common directory simple have a Make.defs on its own and leave boards with their own Makefile?

patacongo commented:
> 
> 
> I realized I must now transition ALL boards under stm32 (not only the ones I modify) to have a Make.defs in src/ instead of a Makefile as the common/ dir's Makefile is now used. It's a bit confusing why this is done so. Couldn't the common directory simple have a Make.defs on its own and leave boards with their own Makefile?

@jerpelea  Alin is the mastermind between the common and shared driver logic under boards/ .  I think we need to engage Alin for advice on most details.

jerpelea commented:
> 
> 
> I realized I must now transition ALL boards under stm32 (not only the ones I modify) to have a Make.defs in src/ instead of a Makefile as the common/ dir's Makefile is now used. It's a bit confusing why this is done so. Couldn't the common directory simple have a Make.defs on its own and leave boards with their own Makefile?

I see that you started the painfull migration and this can only make me happy that we will end up with an better board layout 
I proposed the following layout for the board folder:
common folder - is used for all initialization from all boards that is common 
board_name folder - is used for headers and initialization specific for a board
drivers - specific drivers that can not be used on other boards (Spresense has HW that is not found on other boards and needs drivers for it)

This change has to be applied to all boards that use the same stm32 microcontroller and I would recommend to start in small steps by moving 1 initialization then wait for it to pass the CI and be merged.

For reference please check the cxd56 board

v01d commented:
> I see that you started the painfull migration and this can only make me happy that we will end up with an better board layout
> I proposed the following layout for the board folder:
> common folder - is used for all initialization from all boards that is common
> board_name folder - is used for headers and initialization specific for a board
> drivers - specific drivers that can not be used on other boards (Spresense has HW that is not found on other boards and needs drivers for it)
> 
Painful is a good adjective haha. I moved code to common/ since the board logic to me are not driver per se: it is initialization code plus expected callbacks. Some times this can be interpreted as a driver, but I'm not sure how to draw the line. 
BTW, do you know why the system used to build the common/ folder is based on having a Makefile there which overrides the Makefile in the board directory? It seems that whenever common/ is used, all boards now have to have a Make.defs inside their src/ instead of a Makefile. This is awkward to me but maybe there's a reason for it. Anyways, having one Makefile for board family maybe it is better, but in that case I woud migrate ALL boards to keep consistency (ie, have a common/ directory always there).

> This change has to be applied to all boards that use the same stm32 microcontroller and I would recommend to start in small steps by moving 1 initialization then wait for it to pass the CI and be merged.

I imagine you mean what I just mentioned, I can do that migration, but would like to confirm that one Makefile per board family is the expected layout.

> 
> For reference please check the cxd56 board

Yes, I have been looking at this for reference.

Anyways, what I would like to get feedback on, is the way I'm passing data such as pin configuration, timer number, etc. As I mentioned in #1020 (maybe this PR should be considered in light of that proposal), I don't really like the idea of adding a bunch of structs to hold this data. In #1020 I give an example where this happens.

jerpelea commented:
For our scope, of cleaning the common initialization, I would not encourage you to go the dts way (at least not until we clean all platforms).
Instead I would suggest to get 1 Makefile for each board and all the common initialization in common folder. 
From what I saw on stm32 there is no special driver and I do not think that you should add anything in the drivers for this microcontroller.

v01d commented:
I actually mentioned that because I described there the issue with passing timer/pin information there. Specifically, I wanted to ask you what do you think of this approach:
https://github.com/apache/incubator-nuttx/commit/44972b32ab5dc3c716855fc9f0c4d7781df510e2#diff-42be725a6a20b5b7f6fd1bffc4b96807R38
https://github.com/apache/incubator-nuttx/commit/44972b32ab5dc3c716855fc9f0c4d7781df510e2#diff-dc7f3d864bc967d588bca1dd29e65bb8R147

Is that reasonable? I would need confirmation for that (or a better alternative) before continuing with this PR.

v01d commented:
> Instead I would suggest to get 1 Makefile for each board and move the common initialization in common folder.
Sorry, I got confused, you mean one Makefile per board or per board family (ie, one Makefile under common/). The latter is what is supported in the build system (see https://github.com/apache/incubator-nuttx/blob/46baccb5212660fb671c069bbd569bf56ff47e56/tools/Makefile.unix#L315).

jerpelea commented:
this was my approch on the spresense board
https://github.com/apache/incubator-nuttx/tree/master/boards/arm/cxd56xx/common
https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/common/src/Make.defs
https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/spresense/src/Make.defs

patacongo commented:
> For our scope, of cleaning the common initialization, I would not encourage you to go the dts way (at least not until we clean all platforms).

We also need to get consensus from NuttX users.  In the past, the majority of users were opposed to DTS.  We should confirm that this has become a feature desired by users before investing too much.

v01d commented:
> For the timer/pin values or other board specific definitions I would move define the proper ones in the board.h (for each board).
> @patacongo what do you think about my proposal?

The problem is that it is expected that the board logic includes the common code, not the other way around. So common code could see the board.h, but not the internal boardname.h typically found under the board's src/. So, all this kind of definitions needed by common code should be on board.h and not the internal boardname.h. That sounds reasonable.

v01d commented:
> We also need to get consensus from NuttX users. In the past, the majority of users were opposed to DTS. We should confirm that this has become a feature desired by users before investing too much.
I agree that it is important to have consensus, that is why I opened #1020. We can discuss devicetree there.

v01d commented:
I introduced the change to read config values from board.h. I also converted all boards under stm32 to the Make.defs system. Now all boards under stm32 appear to build correctly.

Going further, I think that it would be useful to start organizing and standardizing these initialization calls. For example, cxd56 does this for sensors as in: https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/common/src/cxd56_sensors.c. There, a `board_sensors_initialize` function is defined which takes care of calling all required initialization functions at once. This mechanism can be generalized to have one `board_<deviceclass>_initialize` call taking care of all calls into initialization functions of that device class. An appropriate location would be something like nuttx/boards/common/board_sensors_initialize.c and so on for different device classes. This would also allow for standardization of board-level initialization calls for each device. For sensors, the following interface can be used (as cxd56xx code expects): board_<sensor>_initialize(<dev number>, <bus number>). After this grouping, board's bringup calls could directly call board_<deviceclass>_initialize(). However, I believe that if this API is standardized it would be more appropriate to have these calls made directly from the OS, as is done currently for board_app_initialize(). In this way, board initialization would really be simplified to only call into board-specific drivers not handled in common directories.

As a first step, I will replicate what cxd56 does and have a stm32_sensors.c under stm32/common (would probably be more or less an exact duplicate). This way I can focus on moving only sensors into common/ and later repeat this for other device classes.

jerpelea commented:
please fix nxstyle errors

v01d commented:
I'm fixing style errors as I go, but it is very time consuming since every file I modify seems to have uncorrected styling problems which I then must fix, so I'm paying dearly for every change. Anyway, this is still a WIP, there's still lots of logic to move into common/.

v01d commented:
I just painstakingly finished moving all sensor initialization logic for stm32 into common/. I tried fixing styling issues but I'm getting a lot of problems with #1031 so I will wait for that to be fixed.

I would like to write a board_sensors_initialize function as explained before but I realized that it would make it impossible to initialize multiple instances of the same sensor, since it is not possible for a board to "request" multiple instances to be created using macros. In fact, that is one of the major reasons for using something like a device tree which would make it possible. While I think no board right now registers more than one instance, it does not feel right to introduce a change that adds this limit so maybe it is better to leave this as is for now (ie: make board logic call each function on its own) and revisit this idea if devicetree is used for this purpose.

Since this PR is becoming very big, maybe once CI passes you would like to merge this and then I can open a PR for each device class in turns. I can improve the organization of files inside common/ and use better naming (board_ prefix for files, not stm32_) to avoid clashes with arch files as the final step here before merge.

jerpelea commented:
I will love to see this one merged after it passes the nxstyles 

For now please do the initialization with idefs in the bringup file (like I did it in the Spresense board) 
https://github.com/apache/incubator-nuttx/blob/master/boards/arm/cxd56xx/spresense/src/cxd56_bringup.c#L417

This way each board "can choose" which sensors are initialized from common

v01d commented:
Ok, done. Haven't reorganized/renamed files yet. I would rather leave that for a future PR when more device classes are handled. 

I think it is important to put down some guidelines down (on the wiki probably) for future board contributions. Enabling support for a sensor should not result in a file inside a board if it can be helped. Instead, the common directory should be used. Moreover, all files inside common have the following interface:
<pre>
board_<sensor>_initialize(<args>)
</pre>
Where `<args>` should be either "int devno" to identify a device (from which the devpath is built) or "int devno, int busno" if it is a SPI/I2C device. For now at least I think it is good enough of a standard.

Ouss4 commented on boards/arm/stm32/common/src/stm32_nunchuck.c:
```diff
@@ -0,0 +1,111 @@
+
+  /* Register the joystick device as /dev/nunchuck0 */
+
+  iinfo("Initialize joystick driver: %s\n", devname);
+
+  snprintf(devpath, 14, "/dev/nunchuck%d", devno);
+  ret = nunchuck_register(devname, i2c);
```
```suggestion
    snprintf(devpath, 15, "/dev/nunchuck%d", devno);
    iinfo("Initialize joystick driver: %s\n", devpath);

  ret = nunchuck_register(devpath, i2c);
```
`devname` is not defined in this scope, it was used as a parameter in the old code.
Also 15 is the minimum to have "/dev/nunchuckX"

v01d replied:
Good catch. The string array also needs to be bigger then. Will change that.

Ouss4 commented on boards/.gitignore:
```diff
@@ -22,3 +22,4 @@ cscope.out
 /*.adb
 /*.lib
 /*.src
+/*/*/common/board
```
Do you need to ignore this directory?

v01d replied:
Yes, it is a symbolic link generated by the build system

Ouss4 commented:
I brought this PR locally and tested it.  It builds and runs fine. (If you want to try networking you'll need to incorporate #1012 changes or rebase this PR).

Ouss4 commented on boards/arm/stm32/common/src/stm32_nunchuck.c:
```diff
@@ -98,10 +98,10 @@ int board_nunchuck_initialize(int devno, int busno)
   /* Register the joystick device as /dev/nunchuck0 */
 
-  iinfo("Initialize joystick driver: %s\n", devname);
+  iinfo("Initialize joystick driver: %s\n", devpath);
```
`devpath` is empty at this point.  This line should be swapped with 103.

v01d replied:
Ok
